crates/bdk/tests/psbt.rs

```diff
@@ -183,7 +183,15 @@ fn test_psbt_multiple_internalkey_signers() {
             },
         )),
     );
-    let _ = wallet.sign(&mut psbt, SignOptions::default()).unwrap();
+    assert!(wallet
+        .sign(
+            &mut psbt,
+            SignOptions {
+                remove_taproot_extras: false,
+                ..Default::default()
+            }
+        )
+        .unwrap());
     // Checks that we signed using the right key
     assert!(
         psbt.finalize_mut(&secp).is_ok(),
```

Comment:
@danielabrozzoni Curiously, removing `tap_key_sig` from a psbt input made this fail somewhere during `finalize_mut`, hence this change. But this shouldn't affect the meaning of the test, other than to assert finalization succeeded in `Wallet::sign`. Just wanted to run this by you because I think it was one of your tests.

Reply:
The reason for this is that `finalize_mut` doesn't really make sense here: `wallet.sign` is already finalizing the transaction (and you are asserting that, thanks for adding it!), and the only reason why `finalize_mut` wouldn't complain before is because we left the taproot_extras in the psbt, and so miniscript was able to re-finalize it. Can you please add a commit removing this assert with finalize_mut? Thanks!

Reply:
Actually, I don't believe asserting that `sign` returns true is sufficient to check *which key* did the signing, whereas `finalize_mut` will take steps to validate sigs, if I understand. If we just have wallet not try to finalize, then this will make more sense.

Reply:
Thank you. How about we take your original suggestion and get rid of `finalize_mut`. 

The only problem I see is that this test (below) passes on the pre #1200 buggy code, when it shouldn't?

<details>

<summary>expand</summary>

```rust
fn test_psbt_multiple_internalkey_signers() {
    use bdk::signer::{SignerContext, SignerOrdering, SignerWrapper};
    use bdk::KeychainKind;
    use bitcoin::{secp256k1::Secp256k1, PrivateKey};
    use miniscript::psbt::PsbtExt;
    use std::sync::Arc;

    let secp = Secp256k1::new();
    let (mut wallet, _) = get_funded_wallet(get_test_tr_single_sig());
    let send_to = wallet.get_address(AddressIndex::New);
    let mut builder = wallet.build_tx();
    builder.add_recipient(send_to.script_pubkey(), 10_000);
    let mut psbt = builder.finish().unwrap();
    // Adds a signer for the wrong internal key, bdk should not use this key to sign
    wallet.add_signer(
        KeychainKind::External,
        // A signerordering lower than 100, bdk will use this signer first
        SignerOrdering(0),
        Arc::new(SignerWrapper::new(
            PrivateKey::from_wif("5J5PZqvCe1uThJ3FZeUUFLCh2FuK9pZhtEK4MzhNmugqTmxCdwE").unwrap(),
            SignerContext::Tap {
                is_internal_key: true,
            },
        )),
    );
    let finalized = wallet.sign(&mut psbt, SignOptions::default()).unwrap();
    assert!(finalized);
    
    // Checks that we signed using the right key
    let err = psbt.finalize_mut(&secp);
    dbg!(err);
}
```

</details>

so `finalize_mut` was doing its job, even though the wallet considers it finalized. So what made the tx unbroadcastable, I wonder? 

This is probably a bit out of scope for the current PR, but in the meantime I'll think of some other ways we can test the 'correct internal key' logic if you think that would be valuable.

Reply:
Oh, you're right, thanks! I think the problem lays in the fact that `finalized` is calculated based on "do we have the descriptor to satisfy this input? If that's so, and we can satisfy every input, then the transaction must be finalized". 

I am a bit torn here. The only solution that comes to my mind is to get rid of the test - we realized it wasn't even testing what I thought it was, and atm we don't have any function to check "did we sign with the right key?". 

@evanlinjin inputs?

Reply:
Edit: we can also comment the test for now instead of removing it, with a comment pointing to this discussion

Reply:
Referring to this

https://github.com/bitcoindevkit/bdk/blob/7aca88474ac241bc0938ff989026463bdaf8f3ab/crates/bdk/src/wallet/signer.rs#L459-L467
 

>  The bug was that we would sign with whatever key we had in hand, thus replacing the previous signature that was in the PSBT
> 
>  In the wallet at the time of signing the key with higher precedence is the right one, so we are signing with the right one first, and the wrong one second.


My interpretation is that we sign with the wrong key first, since it has the lower SignerOrdering, and then skip signing for subsequent signers, as at that point `tap_key_sig` is no longer None.


> After the fix, we check that we're signing with the right key, and so adding a wrong key that signs last doesn't really matter (it is not used for signing anyways)

Agreed

> Edit: we can also comment the test for now instead of removing it, with a comment pointing to this discussion

I'm ok with this.

Reply:
> My interpretation is that we sign with the wrong key first, since it has the lower SignerOrdering, and then skip signing for subsequent signers, as at that point tap_key_sig is no longer None.

Yes, that's exactly what we would do